core/container.go

```diff
@@ -289,30 +289,59 @@ func (container *Container) From(ctx context.Context, addr string) (*Container,
 		return nil, fmt.Errorf("failed to get buildkit client: %w", err)
 	}
 
-	container = container.Clone()
-
 	platform := container.Platform
 
 	refName, err := reference.ParseNormalizedNamed(addr)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to parse image address %s: %w", addr, err)
 	}
+	// add a default :latest if no tag or digest, otherwise this is a no-op
+	refName = reference.TagNameOnly(refName)
 
-	ref := reference.TagNameOnly(refName).String()
+	if refName, isCanonical := refName.(reference.Canonical); isCanonical {
+		return container.FromCanonical(ctx, refName)
+	}
 
-	_, digest, cfgBytes, err := bk.ResolveImageConfig(ctx, ref, sourceresolver.Opt{
+	_, digest, _, err := bk.ResolveImageConfig(ctx, refName.String(), sourceresolver.Opt{
 		Platform: ptr(platform.Spec()),
 		ImageOpt: &sourceresolver.ResolveImageOpt{
 			ResolveMode: llb.ResolveModeDefault.String(),
 		},
 	})
 	if err != nil {
-		return nil, fmt.Errorf("failed to resolve image %s: %w", ref, err)
+		return nil, fmt.Errorf("failed to resolve image %s: %w", refName.String(), err)
+	}
+	canonRefName, err := reference.WithDigest(refName, digest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set digest on image %s: %w", refName.String(), err)
 	}
 
-	digested, err := reference.WithDigest(refName, digest)
+	return container.FromCanonical(ctx, canonRefName)
```

Comment:
Doing this for one codepath, but it's kind of tricky for others, namely the one where:
1. `from` is called with a ref w/out digest and ResolveImageConfig is called to get that digest
2. `from` is re-called with the digest set, ResolveImageConfig is called again to get the actual image config

I think to skip the second call we'd have to make the image config bytes and optional arg to `from`, which feels not super clean. I'm not super concerned about this though; the second call to ResolveImageConfig is prefer local, so it most likely will not actually do anything. Even in the extremely obscure case it does because something got cleaned up, it's not harmful to actually call it again.

Reply:
Yeah, sounds good to me :tada: